testng-core/src/main/java/org/testng/internal/invokers/MethodInvocationHelper.java

```diff
@@ -267,6 +269,13 @@ public void runTestMethod(ITestResult tr) {
     if (error[0] != null) {
       throw error[0];
     }
+    if (!callBackInvoked.get()) {
+      throw new TestNGException(
+          "IHookable callback was defined but never called for "
```

Comment:
I think defining a callback and not calling it is a valid usage.
Instead, we can enforce that the user must set a status on the result, fallback to skipped + log, or throw an exception.

Reply:
@juherr can you give me an example of when it would be a valid use case ? Also you mentioned that the status need to be set and/or exception thrown. Can you please elaborate on how testng can enforce a status on a test that wasnt invoked and what would that status be ? I didnt quite understand this.

Reply:
In fact, it is a good hook if you want any condition on the run of a test.

Reply:
If you set the result status to STARTED, you'll be able to detect if the status was changed by the user.

Reply:
@juherr - So if I understand this correctly I am guessing that a user deliberately skipping invoking the callback can be equated to the user wanting to skip the test. So what if we threw a skip exception ?

Reply:
> So what if we threw a skip exception ?

Who? Where?
SkipException is for the userland and we try to avoid exceptions in listeners ;)

Reply:
Who -> TestNG
Where -> From within TestNG

> SkipException is for the userland and we try to avoid exceptions in listeners ;)

Not necessarily true. TestNG could also throw this exception to signal that in this instance it was TestNG that skipped executing something (And this is the ONLY way in which TestNG marks something as eligible for being skipped) 

And no we are not throwing exceptions in listeners. We would be throwing the exception from within `MethodInvocationHelper` which is not a listener

Reply:
By default, I don't like to use exceptions for legit workflow (instead I prefer the functional way like Either).
But I admit they are helpful.

When possible, I'd like to avoid them in the core, especially SkipException which is somehow part of the api.